### ISSUE TITLE
Make it possible to query a type literal/class reference

### ISSUE THREAD

**Issue by eEQK** [dart-lang/macros#156]
_TL;DR do we want a `ClassReference.uri` field? or maybe `ClassReference.name` to be a `QualifiedName`?_

---

I'm currently building a macro that will take in a specification as an argument and generate a client/server based on this specification

the way it's going to be used will be something along the lines of:
```dart
@Client(BaseService)
class BaseClient {}
```

which generates the following:
```json
{
  "type": "ConstructorInvocation",
  "value": {
    "type": { ... },
    "constructor": { ... },
    "arguments": [
      {
        "type": "PositionalArgument",
        "value": {
          "expression": {
            "type": "TypeLiteral",
            "value": {
              "typeAnnotation": {
                "type": "NamedTypeAnnotation",
                "value": {
                  "reference": {
                    "type": "ClassReference",
                    "value": {
                      "name": "BaseService"
                    }
                  },
                  "typeArguments": []
                }
              }
            }
          }
        }
      }
    ]
  }
}
```

As you can see, having a name only is not enough to query `BaseService` since at that moment `BaseService` is not loaded/resolved (`builder.model.node['types']`)

```json
"types": {
    "named": {
      "dart:core#Object": { ... },
      "dart:core#Null": { ... },
      "dart:async#Future": { ... },
      "package:bar/test.dart#BaseClient": { ... }
  }
```

I do actually have a local fix for that, that adds `uri` to `ClassReference` (I'm not sure but I feel like it makes the most sense to put that information there):

```json
                  "reference": {
                    "type": "ClassReference",
                    "value": {
                      "name": "BaseService",
                      "uri": "package:bar/test.dart"
                    }
                  },
```

and I'd love to submit a PR, but I'm not sure if that's how the model should look like? another approach would be to make `name` a qualified name, like so:

```
                  "reference": {
                    "type": "ClassReference",
                    "value": {
                      "name": "package:bar/test.dart#BaseService"
                    }
                  },
```

So my question is, do we want a `uri` field, `name` to become a QualifiedName or maybe something else?

Thanks in advance!

**Comment by davidmorgan** [comment]
Thanks for filing.

I think we're not ready to get into this level of detail here yet, there are bigger questions about the macro metadata model that we need to figure out first.
